prow/tide/tide.go

```diff
@@ -1151,6 +1155,15 @@ type PullRequest struct {
 	Milestone *struct {
 		Title githubql.String
 	}
+	Reviews struct {
+		Nodes []Review
+	} `graphql:"reviews(states: APPROVED, last: 100)"`
```

Comment:
100 seems safe, but "we'll never have more than 100 approvers on a PR" is sort of like "we'll never need [dates > 2038][1]".  I'm happy to work up something more robust here if folks are concerned about hitting this limit.

[1]: https://en.wikipedia.org/wiki/Year_2038_problem

Reply:
There is no guarantee that approvers leave an approving review, vs a comment on the pr with an `/lgtm` and/or `/approve` command.

Reply:
> There is no guarantee that approvers leave an approving review, vs a comment on the pr with an `/lgtm` and/or `/approve` command.

Yup.  From my initial PR comment, see the "This is pulling approvers from GitHub because we can do that with straightforward GraphQL..." paragraph ;).  Another approach would be to have the `lgtm` and `approve` plugins push GitHub approvals if the author did not already have one in place.  But for a first pass at this opt-in feature, I think just pulling GitHub approvers is fine (unless maintainers need me to do more to get it accepted ;).

Reply:
Are you up for trying to get the list of people who approved? In either this commit or a follow-up? This would be a good feature to enable across the whole kubernetes org if so.

/cc @cblecker

Reply:
> Are you up for trying to get the list of people who approved?

I'm fine with listing folks who filed GitHub approvals despite not being authorized for `/lgtm` or `/approve`.  Are you on board with that?

It would be nice to *also* include folks who use `/lgtm` or `/approve` but don't explicitly file a GitHub approval.  The bot probably doesn't have the ability to push approvals to GitHub on their behalf.

Maybe the `lgtm` and `approve` plugins could push a special bot approval?  The downside is that they'd have to coordinate on that approval message.

Or the Tide plugin would have to call out to the `lgtm` and `approve` plugin logic (for repos where the plugins were enabled) and take another round of parsing to look for missing approvers?

Or is there a way for the `hook` command to pass information between plugins?  Actually, the `lgtm` and `approve` plugins are probably just reacting to individual events; do they even track the full state of the pull request?

Anyhow, I'm happy to work on getting `/lgtm` and `/approve` supported here, but I don't understand the moving pieces well enough to have a feeling for the best approach.

Reply:
I would rather tide get the list of approvals from the approve plugin, than query GitHub at all. That way, the list is consistent between the comment that the approve plugin leaves saying `This pull-request has been approved by: fejta, wking` for example. Less surprises, because the forward facing list matches the in-the-backround list. If the `review_acts_as_approve` is enabled, then the functionality would match.

Reply:
Yes, anyone can /approve... they just might happen to approve 0 of the files modified in the PR so including them is fine 😀 

Although if you only want to count people who count, you could use this method: https://github.com/kubernetes/test-infra/blob/master/prow/plugins/approve/approvers/owners.go#L59


Right, the bot can't push a github approval on their behalf.

The most straight forward approach would be to scan through all the issue comments (and review comments) in the PR and add the person to the list whenever the leave a /lgtm and/or /approve comment and removing them whenever they leave a `/lgtm cancel` or `/approve cancel` comment.

Maybe refactor this method: https://github.com/kubernetes/test-infra/blob/master/prow/plugins/approve/approve.go#L539

Reply:
> Yes, anyone can `/approve...` they just might happen to approve 0 of the files modified in the PR so including them is fine

This raises issues around rebase-style merging with multiple commits.  What about folks who `/approve` after reviewing the files they're responsible for but without looking at the other changes.  Then we merge the PR, and add them as reviewers to all of the rebased commits (I think this is how the GitHub merge API will handle this case; but I haven't tested that).  Is that acceptable?  Or do we need to find a way to do a rebase merge with commit-specific appended text?

Reply:
> If the `review_acts_as_approve` is enabled, then the functionality would match.

This setting is configured per-repo group.  If we need to port to using `approve` logic to calculate the approver set to get this PR landed (vs. pulling it with GraphQL from the PR approver set like I'm doing as of 6f8a4f423b), do you also want me to make `merge_approve_tag` more granular than per-Tide (as discussed in the initial PR comment)?  If so, since we're already talking about using the `approve` logic for this approach, we might be able to move `merge_approve_tag` into the `Approve` config (as a sibling of `review_acts_as_approve`), and have the `approve` logic that spits out approvers just return an empty set if `review_acts_as_approve` was empty.

On the other hand, I don't know how to square that with also pulling the approver set from the `lgtm` plugin for repos where the `approve` plugin is not enabled at all.  And what if the `approve` plugin is enabled, but for some weird reason, the repo is configured to only require the `lgtm` to gate merges?  The current coupling between `lgtm`/`approve` and Tide is very weak (just the labels?), and all the ideas for coupling them more tightly together seem brittle.  My personal preference is still to just pull GitHub approvers from the PR with GraphQL like I do now, and not worry about syncing with the `lgtm` and `approve` plugins :p.

Reply:
The issue is, GitHub approvals are meaningless to us (unless enabled in the approve/lgtm plugins). So this would add meaningless information to the merge commit if we just left it at that.

Reply:
Right, but in the context of tide, GitHub-approving a PR is meaningless to tide. I don't think we should duck that fact at all. What's the point in recording who approved a PR for merge if it doesn't actually reflect who approved a PR for merge? The first sentence of your PR body outlines this "Make it easier to find out who approved a given commit/merge by appending that information to the commit message.".

Reply:
> Right, but in the context of tide, GitHub-approving a PR is meaningless to tide.

And in the context of Tide, the `/approve` and `/lgtm` actions are also currently meaningless.  They're only meaningful because:

1. Users often configure Tide to require those labels (e.g. [here][1]).
2. User often configure the `lgtm` and `approve` plugins to set those labels (e.g. [here][2] and [here][3]).

So tying Tide to that channel is currently as much of a leap as trying Tide to GitHub approval.

> What's the point in recording who approved a PR for merge if it doesn't actually reflect who approved a PR for merge?

Because as a reviewer I can add GitHub approvals.  While without this PR, I can't do anything to get my review annotated in the commit message.

Would it help if I update the config parameter to `merge_github_approve_tag`?  Or are we really committed to tightly binding Tide to the `approve` (and `lgtm`?) plugins?  If it's the latter, it seems like we'd want to rework the Tide/`approve` integration, and then come back and revisit commit message tags once we'd got them more tightly connected.

[1]: https://github.com/kubernetes/test-infra/blob/ff508f632fbe5847cad1fb04e3a08d2c92b63b83/prow/config.yaml#L355-L356
[2]: https://github.com/kubernetes/test-infra/blob/ff508f632fbe5847cad1fb04e3a08d2c92b63b83/prow/plugins.yaml#L66
[3]: https://github.com/kubernetes/test-infra/blob/ff508f632fbe5847cad1fb04e3a08d2c92b63b83/prow/plugins.yaml#L96

Reply:
I think that annotating commits with misleading information is probably not the right path to go down. You're right in that tide doesn't care who "approves" the merge, as long as the PR meets the merge requirements.

The other way to keep things more loosely linked would be somehow allowing plugins to specify annotations for merge, and then have tide pull in and combine all those annotations at the merge stage. Then tide wouldn't really care at all what was in the specific annotations, but the plugins could handle that.

Either way, the point that I'm feeling more and more strongly about the more I think about it is that recorded information in git history should be accurate. If we are recording approvals (which I think is a fantastic idea btw), we should be recording accurate approvals that will actually action the merge.

Reply:
> I think that annotating commits with misleading information is probably not the right path to go down.

It's not misleading if we rename the setting to `merge_github_approve_tag` and folks configure it to use a `GitHub-Approve` tag, is it?

> The other way to keep things more loosely linked would be somehow allowing plugins to specify annotations for merge, and then have tide pull in and combine all those annotations at the merge stage. Then tide wouldn't really care at all what was in the specific annotations, but the plugins could handle that.

That sounds workable.  I'm not sure how much work it would be, though ;).

Reply:
> The other way to keep things more loosely linked would be somehow allowing plugins to specify annotations for merge...

Thinking through this some more, where's the balance between query efficiency vs. Go isolation?  Would the `approve` and `lgtm` plugins conspire to share a single PR-comment request per about-to-be-merged PR (easier on API quotas)?  Or would they stay isolated and each make their own query (easier on test-infra devs)?

Reply:
> It's not misleading if we rename the setting to `merge_github_approve_tag` and folks configure it to use a `GitHub-Approve` tag, is it?

I've shifted to `merge_github_approve_tag` and updated the recommended value to `GitHub-approved-by` in b0c751d -> 97f6fa5.  That hopefully makes the PR in its current state less contentious while we work out if/how we want to handle plugin-injected commit tags.

Reply:
Somewhat related: it's entirely possible to use tide without using our approval mechanisms *or github's built-in mechanisms* at all, we should document this feature clearly to note what modes of approval it actually supports when we settle on those :-)

Reply:
> ... we should document this feature clearly to note what modes of approval it actually supports...

Done in 97f6fa588d8 [here][1] and [here][2].  If/when we update the approach to use different logic, I'll update those docs too.

[1]: https://github.com/kubernetes/test-infra/commit/97f6fa588d898a1a45cf2b0d7373e25292bdafad#diff-7becb33d12040460d59c609e7db2e02cR18
[2]: https://github.com/kubernetes/test-infra/commit/97f6fa588d898a1a45cf2b0d7373e25292bdafad#diff-1263895d4391f05abbce7f1a78041321R84

Reply:
(also I would agree that we should try to record prow's typical approval mechanism, most of our approvals are not via github approve, github's ACL models are generally insufficient for kubernetes).